pkg/assetfetcher/http.go

```diff
@@ -58,57 +57,62 @@ type NexposeAssetFetcher struct {
-	errChan := make(chan error)
-	assetChan := make(chan domain.Asset)
-	go func() {
-		var wg sync.WaitGroup
-		defer close(assetChan)
-		defer close(errChan)
-		// make the first call to Nexpose to get the total number of pages we'll
-		// have to page through to get all the assets, start with page 0
-		var currentPage = 0
-		req, err := newNexposeSiteAssetsRequest(c.Host, siteID, currentPage, c.PageSize)
-		if err != nil {
-			errChan <- err
-			return
-		}
-
-		res, err := http.DefaultClient.Do(req.WithContext(ctx))
-		if err != nil {
-			errChan <- &NexposeHTTPRequestError{err, req.URL.String()}
-			return
-		}
-		defer res.Body.Close()
-		respBody, err := ioutil.ReadAll(res.Body)
-		if err != nil {
-			errChan <- &ErrorReadingNexposeResponse{err, req.URL.String()}
-			return
-		}
-
-		var siteAssetResp SiteAssetsResponse
-		if err := json.Unmarshal(respBody, &siteAssetResp); err != nil {
-			errChan <- &ErrorParsingJSONResponse{err, req.URL.String()}
-			return
-		}
-
-		for _, asset := range siteAssetResp.Resources {
-			assetChan <- asset
-		}
-
-		// We've gotten the first page (page 0) and added the assets to the channel,
-		// so here we'll increment the currentPage to start on page 1 and use TotalPages, which is
-		// the total number of pages of assets that Nexpose has, to paginate through to get all the assets
+	errChan := make(chan error, 1)
+	defer close(errChan)
+
+	// have to page through to get all the assets, start with page 0
+	// make the first call to Nexpose to get the total number of pages we'll
+	var currentPage = 0
+	req, err := newNexposeSiteAssetsRequest(c.Host, siteID, currentPage, c.PageSize)
+	if err != nil {
+		errChan <- err
+		return nil, errChan
+	}
+
+	res, err := http.DefaultClient.Do(req.WithContext(ctx))
+	if err != nil {
+		errChan <- &NexposeHTTPRequestError{err, req.URL.String()}
+		return nil, errChan
+	}
+	defer res.Body.Close()
+	respBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		errChan <- &ErrorReadingNexposeResponse{err, req.URL.String()}
+		return nil, errChan
+	}
+
+	var siteAssetResp SiteAssetsResponse
+	if err := json.Unmarshal(respBody, &siteAssetResp); err != nil {
+		errChan <- &ErrorParsingJSONResponse{err, req.URL.String()}
+		return nil, errChan
+	}
+
+	pagedAssetChan := make(chan domain.Asset, siteAssetResp.Page.TotalResources)
+	pagedErrChan := make(chan error, siteAssetResp.Page.TotalResources)
+
+	for _, asset := range siteAssetResp.Resources {
+		pagedAssetChan <- asset
+	}
+
+	// We've gotten the first page (page 0) and added the assets to the channel,
+	// so here we'll increment the currentPage to start on page 1 and use TotalPages, which is
+	// the total number of pages of assets that Nexpose has, to paginate through to get all the assets
+	currentPage++
+	totalPages := siteAssetResp.Page.TotalPages
+	var wg sync.WaitGroup
+	for currentPage < totalPages {
```

Comment:
nit: a while loop with an increment is usually equivalent to a for loop. For example, this code reads:

```golang
currentPage++
while currentPage < totalPages {
  currentPage++
}
```
which is equivalent to
```golang
for currentPage := 1; currentPage < totalPages; currentPage++ { // or currentPage += 1 or even currentPage = currentPage + 1 (my fave)
}
```

Generally speaking, I think the second is better when we can use it because all of the state management is documented on one line instead of throughout the loop. It's not a blocker to anything though. More of a preference.

Reply:
I'll buy it